src/ring_theory/hahn_series.lean

```diff
@@ -382,14 +406,134 @@ noncomputable instance [comm_semiring R] : comm_semiring (hahn_series Γ R) :=
   end,
   .. hahn_series.semiring }
 
-noncomputable instance [ring R] : ring (hahn_series Γ R) :=
+instance [ring R] : ring (hahn_series Γ R) :=
 { .. hahn_series.semiring,
   .. hahn_series.add_comm_group }
 
-noncomputable instance [comm_ring R] : comm_ring (hahn_series Γ R) :=
+instance [comm_ring R] : comm_ring (hahn_series Γ R) :=
 { .. hahn_series.comm_semiring,
   .. hahn_series.ring }
 
+section semiring
+variables [semiring R]
+
+@[simp]
+lemma single_mul_single {a b : Γ} {r s : R} :
+  single a r * single b s = single (a + b) (r * s) :=
+begin
+  ext x,
+  by_cases h : x = a + b,
+  { rw [h, mul_single_coeff_add],
+    simp },
+  { rw [single_coeff_of_ne h, mul_coeff, finset.sum_eq_zero],
+    rintros ⟨y1, y2⟩ hy,
+    obtain ⟨rfl, hy1, hy2⟩ := finset.mem_add_antidiagonal.1 hy,
+    rw [eq_of_mem_support_single hy1, eq_of_mem_support_single hy2] at h,
+    exact (h rfl).elim }
+end
+
+/-- -/
+def C : R →+* (hahn_series Γ R) :=
+{ to_fun := single 0,
+  map_zero' := single_eq_zero,
+  map_one' := rfl,
+  map_add' := λ x y, by { ext a, by_cases h : a = 0; simp [h] },
+  map_mul' := λ x y, by rw [single_mul_single, zero_add] }
+
+@[simp]
+lemma C_apply (r : R) : C r = single (0 : Γ) r := rfl
+
+@[simp]
+lemma C_zero : C (0 : R) = (0 : hahn_series Γ R) := C.map_zero
+
+@[simp]
+lemma C_one : C (1 : R) = (1 : hahn_series Γ R) := C.map_one
+
+lemma C_mul_eq_smul {r : R} {x : hahn_series Γ R} : C r * x = r • x :=
+single_zero_mul_eq_smul
+
+end semiring
+
+section algebra
+variables [comm_ring R] {A : Type*} [semiring A] [algebra R A]
+
+instance : algebra R (hahn_series Γ A) :=
```

Comment:
Can you add an `instance [is_scalar_tower R S A] : is_scalar_tower R S (hahn_series Γ A)` (+ probably some extra assumptions) instance too?

Reply:
Although on second thoughts, `is_scalar_tower` and `smul_comm_class` belong up near `semimodule`, not by `algebra`, so perhaps this is out of scope.